src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts

```diff
@@ -0,0 +1,509 @@
+
+const contractAddresses: Partial<
+  Record<
+    Chain,
+    { carbonControllerAddress: string; voucherContractAddress: string }
+  >
+> = {
+  [Chain.Ethereum]: {
+    carbonControllerAddress: '0xC537e898CD774e2dCBa3B14Ea6f34C93d5eA45e1',
+    voucherContractAddress: '0x3660F04B79751e31128f6378eAC70807e38f554E',
+  },
+}
+
+enum StrategyUpdateReason {
+  Edit = '0',
+  Trade = '1',
+}
+
+export class CarbonDeFiStrategiesAdapter extends SimplePoolAdapter {
+  productId = 'strategies'
+  protocolId: Protocol
+  chainId: Chain
+
+  adaptersController: AdaptersController
+
+  provider: CustomJsonRpcProvider
+
+  constructor({
+    provider,
+    chainId,
+    protocolId,
+    adaptersController,
+  }: ProtocolAdapterParams) {
+    super({
+      provider,
+      chainId,
+      protocolId,
+      adaptersController,
+    })
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+  }
+
+  protected async fetchProtocolTokenMetadata(
+    _protocolTokenAddress: string,
+  ): Promise<Erc20Metadata> {
+    throw new NotImplementedError()
+  }
+
+  protected async fetchUnderlyingTokensMetadata(
+    _protocolTokenAddress: string,
+  ): Promise<Erc20Metadata[]> {
+    throw new NotImplementedError()
+  }
+
+  protected async getUnderlyingTokenBalances(_input: {
+    userAddress: string
+    protocolTokenBalance: TokenBalance
+    blockNumber?: number
+  }): Promise<Underlying[]> {
+    throw new NotImplementedError()
+  }
+  protected async getUnderlyingTokenConversionRate(
+    _protocolTokenMetadata: Erc20Metadata,
+    _blockNumber?: number | undefined,
+  ): Promise<UnderlyingTokenRate[]> {
+    throw new NotImplementedError()
+  }
+
+  /**
+   * Update me.
+   * Add your protocol details
+   */
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'CarbonDeFi',
+      description: 'CarbonDeFi adapter for strategy balances',
+      siteUrl: 'https://carbondefi.xyz',
+      iconUrl: 'https://app.carbondefi.xyz/favicon.ico',
+      positionType: PositionType.Supply,
+      chainId: this.chainId,
+      productId: this.productId,
+    }
+  }
+
+  /**
+   * Update me.
+   * Returning an array of your protocol tokens.
+   */
+  async getProtocolTokens(): Promise<Erc20Metadata[]> {
+    throw new NotImplementedError()
+  }
+
+  /**
+   * Update me.
+   * Add logic to get userAddress positions in your protocol
+   */
+  async getPositions({
+    userAddress,
+    blockNumber,
+  }: GetPositionsInput): Promise<ProtocolPosition[]> {
+    const carbonControllerContract = CarbonController__factory.connect(
+      contractAddresses[this.chainId]!.carbonControllerAddress,
+      this.provider,
+    )
+    const voucherContract = Voucher__factory.connect(
+      contractAddresses[this.chainId]!.voucherContractAddress,
+      this.provider,
+    )
+
+    const strategyIds: BigNumberish[] = await voucherContract.tokensByOwner(
```

Comment:
SDK has getUserStrategies - not relevant?

Reply:
As per their Contributing instructions:
https://github.com/bancorprotocol/mmi-defi-adapters/blob/9a4e279f493be26aa4a1cd9242348d852fca1020/CONTRIBUTING.md?plain=1#L26